practice/prototype_anc_samp.py

```diff
@@ -0,0 +1,627 @@
+            # already indexed to be after the first wave of generation (at population size 2 * args.popsize)
+            # so just need to offset by the number of additional coalescent nodes
+            node_offset = self.nodes.num_rows - 2 * args.popsize
+        else:
+            dt = generation - self.last_gc_time
+            self.nodes.set_columns(flags=self.nodes.flags,
+                                   population=self.nodes.population,
+                                   time=self.nodes.time + dt)
+
+        # Create "flags" for new nodes.
+        # This is much faster than making a list
+        flags = np.empty([len(tracker.nodes)], dtype=np.uint32)
+        flags.fill(1)
+
+        # Convert time from forwards to backwards
+        tracker.convert_time()
+        meta_list = [Meta(mut['position'], mut['origin_generation'])
+                     for mut in tracker.mutations]
+        encoded, offset = msprime.pack_bytes(
+            list(map(pickle.dumps, meta_list)))
+
+       # Update internal *Tables
+        self.nodes.append_columns(flags=flags,
+                                  population=tracker.nodes['population'],
+                                  time=tracker.nodes['generation'])
+        self.edges.append_columns(left=tracker.edges['left'],
+                                  right=tracker.edges['right'],
+                                  # only offset mutation and child node ids (and sample ids), for a single simulation generation, edge parents will be correct
+                                  parent=tracker.edges['parent'],
+                                  child=tracker.edges['child'] + node_offset)
+        self.sites.append_columns(position=tracker.mutations['position'],
+                                  ancestral_state=np.zeros(
+                                      len(tracker.mutations['position']), np.int8) + ord('0'),
+                                  ancestral_state_offset=np.arange(len(tracker.mutations['position']) + 1, dtype=np.uint32))
+        self.mutations.append_columns(site=np.arange(len(tracker.mutations['node_id']), dtype=np.int32) + self.mutations.num_rows,
+                                      node=tracker.mutations['node_id'] +
+                                      node_offset,
+                                      derived_state=np.ones(
+                                          len(tracker.mutations['node_id']), np.int8) + ord('0'),
+                                      derived_state_offset=np.arange(
+                                          len(tracker.mutations['position']) + 1, dtype=np.uint32),
+                                      metadata_offset=offset, metadata=encoded)
+        # Sort and simplify
+        msprime.sort_tables(nodes=self.nodes, edges=self.edges,
+                            sites=self.sites, mutations=self.mutations)
+        all_samples = list(set((tracker.anc_samples + node_offset).tolist() +
```

Comment:
It'd be helpful to comment here that the use of `set(...)` guards against entering a node in twice as a sample if the gc generation == the ancient sample generation.

Reply:
done

"# set guards against duplicate node ids when an ancestral sample generation overlaps with a gc generation"